python-package/lightgbm/engine.py

```diff
@@ -486,12 +492,17 @@ def cv(params, train_set, num_boost_round=100,
     """
     if not isinstance(train_set, Dataset):
         raise TypeError("Training only accepts Dataset object")
-
     params = copy.deepcopy(params)
-    if fobj is not None:
-        for obj_alias in _ConfigAliases.get("objective"):
-            params.pop(obj_alias, None)
-        params['objective'] = 'none'
+    if isinstance(params.get('objective'), str) and callable(fobj):
+        params['objective'] = fobj
```

Comment:
The GitHub UI is not allowing me to add a new comment under https://github.com/microsoft/LightGBM/pull/5052#discussion_r830513129, so I'll continue that thread over here on the relevant code.

> @jameslamb there's one caveat with _choose_param_value, when objective is string and fobj callable, the second one should take precedence

I see you're referring to the code below, where if `fobj` is not `None`, it is preferred to anything found in `params`.

https://github.com/microsoft/LightGBM/blob/8e721c564d4534820a7b7febf81382ccf247b5c9/python-package/lightgbm/engine.py#L127-L131

I personally support making the breaking change in this PR that if `fobj` is not `None` and if a value for `objective` or its aliases is found in `params`, the value from `params` is preferred. I think that across all places in the Python and R packages, LightGBM should try to provide the behavior "values provided via `params` override keyword arguments".

Let's get others' opinions before making changes though.

-------


@StrikerRUS @shiyu1994 @guolinke  what do you think? Given the following:

```python
lgb.train(
    params={"objective": _other_custom_objective_function},
    fobj=_some_custom_objective_function,
    ...
)
```

- **option 1**: use the function passed to `fobj` (current behavior)
- **option 2**: use the function passed to `params` (breaking change)
- **option 3**: raise an error about conflicting objectives (breaking change)
- **option 4**: something else

My preference is for **option 2**, for consistency with how `params` is used in most other places in the Python and R packages.

Reply:
Oh, I'm sorry, I hadn't noticed this comment before writing my question here https://github.com/microsoft/LightGBM/pull/5052#discussion_r830537218.
I think we can remove `fjob` argument and work only with `objective` parameter (and its' aliases).

Reply:
Sorry to split it somewhere else. I don't know why GitHub wasn't letting me add a new comment in that other thread (it seems like I can now).

Let's resolve this and continue the conversation in that other thread.

Reply:
Yeah, I noticed that GitHub behaves differently (oddly) depending on where you're trying to write comments: on the `Conversation` or `Files changed` tab.

![image](https://user-images.githubusercontent.com/25141164/159142839-af5fb2cb-e58b-4d68-91d9-629b7c646e33.png)